### PR TITLE
upsample stage (2x)

### PR DESCRIPTION
**Pull request by mo271**

### CODE REVIEW

File: jxl/src/headers/transform_data.rs [hunk 1]
```diff
@@ -324,6 +324,7 @@ const DEFAULT_KERN_8: [f32; 210] = [
 #[nonserialized(CustomTransformDataNonserialized)]
 pub struct CustomTransformData {
     #[all_default]
+    #[default(true)]
```
Comment (veluca93) [inline jxl/src/headers/transform_data.rs:327]: You probably need to sync :-)

File: jxl/src/render/stages/chroma_upsample.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+use crate::render::{RenderPipelineInOutStage, RenderPipelineStage};
+
+pub struct NearestNeighbourUpsample {
```
Comment (veluca93) [inline jxl/src/render/stages/chroma_upsample.rs:8]: This doesn't *quite* fit here anymore if we call it chroma_upsample...

File: jxl/src/render/stages/upsample.rs [hunk 3]
```diff
@@ -3,135 +3,80 @@
...
-            let scaled_cur = input[0][i + 1] * 0.75;
-            let prev = input[0][i];
-            let next = input[0][i + 2];
-            let left = 0.25 * prev + scaled_cur;
-            let right = 0.25 * next + scaled_cur;
-            output[0][2 * i] = left;
-            output[0][2 * i + 1] = right;
+impl Upsample2x {
+    pub fn new(ups_factors: &CustomTransformData, channel: usize) -> Upsample2x {
+        // don't copy
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:21]: not sure I understand this comment...

File: jxl/src/headers/transform_data.rs (hunk 1 above)
Comment (mo271) [reply jxl/src/headers/transform_data.rs]: done

File: jxl/src/render/stages/upsample.rs [hunk 4]
```diff
@@ -3,135 +3,81 @@
...
-
-pub struct VerticalChromaUpsample {
-    channel: usize,
-}
-
-impl VerticalChromaUpsample {
-    pub fn new(channel: usize) -> VerticalChromaUpsample {
-        VerticalChromaUpsample { channel }
+    pub fn kernel(&self, x: usize, y: usize, ix: usize, iy: usize) -> f32 {
+        self.kernel[0][0][if y % 2 == 0 { iy } else { 4 - iy }]
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:38]: mild preference for moving those to variables :-)

File: jxl/src/render/stages/upsample.rs (hunk 3 above)
Comment (mo271) [reply jxl/src/render/stages/upsample.rs:21]: was a leftover....removed.

File: jxl/src/render/stages/upsample.rs [hunk 5]
```diff
@@ -140,88 +86,40 @@ impl RenderPipelineStage for VerticalChromaUpsample {
...
-
-    #[test]
-    fn vchr_consistency() -> Result<()> {
-        crate::render::test::test_stage_consistency::<_, f32, f32>(
-            VerticalChromaUpsample::new(0),
-            (500, 500),
-            1,
-        )
-    }
+            .row(3)
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:96]: Why not .row(3)[3] = 1.0?

File: jxl/src/render/stages/upsample.rs [hunk 6]
```diff
@@ -140,88 +86,40 @@ impl RenderPipelineStage for VerticalChromaUpsample {
...
+                assert_eq!(output[0].as_rect().row(j)[i], 0.0);
+                assert_eq!(output[0].as_rect().row(i)[j], 0.0);
+                assert_eq!(output[0].as_rect().row(13 - j)[i], 0.0);
+                assert_eq!(output[0].as_rect().row(i)[13 - j], 0.0);
+            }
+        }
+        // Check the kernel is copied by the "1.0" in the middle
+        for i in 2..12 {
+            for j in 2..12 {
+                // TODO(firsching): check more specifically what weights maps to what output. Also don't perhaps use equality of floats here.
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:115]: +1 on this ;)

File: jxl/src/render/stages/upsample.rs [hunk 7]
```diff
@@ -140,88 +86,40 @@ impl RenderPipelineStage for VerticalChromaUpsample {
...
-            HorizontalChromaUpsample::new(0),
-            (500, 500),
-            1,
-        )
-    }
-
-    #[test]
-    fn test_hchr() -> Result<()> {
-        let mut input = Image::new((3, 1))?;
+    fn test_upsample2() -> Result<()> {
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:91]: Can we also test:
- consistency
- upsampling a constant image
?

File: jxl/src/render/stages/chroma_upsample.rs (hunk 2 above)
Comment (mo271) [reply jxl/src/render/stages/chroma_upsample.rs:8]: ok, let's move nearest neighbor to just upsample.rs? or should we make a file nearest_neighbor.rs?

File: jxl/src/render/stages/upsample.rs (hunk 7 above)
Comment (mo271) [reply jxl/src/render/stages/upsample.rs:120]: yes, added those

File: jxl/src/render/stages/upsample.rs (hunk 5 above)
Comment (mo271) [reply jxl/src/render/stages/upsample.rs]: done

File: jxl/src/render/stages/upsample.rs [hunk 8]
```diff
@@ -3,135 +3,82 @@
...
-impl RenderPipelineStage for VerticalChromaUpsample {
-    type Type = RenderPipelineInOutStage<f32, f32, 0, 1, 0, 1>;
+impl RenderPipelineStage for Upsample2x {
+    type Type = RenderPipelineInOutStage<f32, f32, 2, 2, 1, 1>;
 
     fn uses_channel(&self, c: usize) -> bool {
         c == self.channel
     }
-
+    // Takes a 5x5 area and upsamples it to 10x10 using the kernel.
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:57]: Not quite a 5x5 area, right?
(also, I assume this was meant to be a doc comment?)

**Review by veluca93 — APPROVED** [review]

File: jxl/src/render/stages/upsample.rs (hunk 8 above)
Comment (mo271) [reply jxl/src/render/stages/upsample.rs]: made it a doc comment and made it more precise

**Pull request merged**